File: Backend/src/main/java/com/example/easynotes/model/HoaDon.java
package com.example.easynotes.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.Nationalized;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import javax.persistence.*;
import java.io.Serializable;
import java.util.Date;
import java.util.List;
import java.util.concurrent.atomic.AtomicReference;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
@EntityListeners(AuditingEntityListener.class)
@JsonIgnoreProperties(value = {"createdAt", "updatedAt"}, allowGetters = true)
@Entity
@Table(name = HoaDon.TABLE_NAME)
public class HoaDon implements Serializable {
    public static final String TABLE_NAME= "HoaDon";

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long maHoaDon;

    @ManyToOne
    @JoinColumn(name = "maNhanVien")
    private NhanVien nguoiLapHoaDon;

    @Column(name = "daXoa")
    private boolean isDeleted;

    @OneToOne
    @JoinColumn(name = "maThanhToan")
    private ThanhToan thanhToan;

    @JsonIgnore
    @OneToMany(mappedBy = "hoaDon", cascade = CascadeType.ALL, fetch = FetchType.EAGER)
    private List<ChiTietHoaDon> chiTietHoaDons;

    @ManyToOne
    @JoinColumn(name = "maSoBan")
    private Ban ban;

    @Nationalized
    @Column(name = "ghiChu")
    private String ghiChu;

    @Column(name = "daTraHoaDon")
    private boolean daTraHoaDon;

    @Enumerated(EnumType.STRING)
    @Column(name = "tinhTrang")
    private TinhTrang tinhTrang;

    @Column(nullable = false, updatable = false, name = "ngayTao")
    @Temporal(TemporalType.TIMESTAMP)
    @CreatedDate
    private Date createdAt;

    @Column(nullable = false, name = "ngayCapNhat")
    @Temporal(TemporalType.TIMESTAMP)
    @LastModifiedDate
    private Date updatedAt;

    public Double tongThanhTien(){
        AtomicReference<Double> tongTien = new AtomicReference<>((double) 0);
        this.chiTietHoaDons.forEach(x -> {
            tongTien.set(tongTien.get() + x.getThanhTien());
        });
        return tongTien.get();
    }
}

File: App/go_quick_app/lib/models/hoa_don.dart
// To parse this JSON data, do
//
//     final hoaDon = hoaDonFromJson(jsonString);

import 'package:go_quick_app/models/ban.dart';
import 'package:go_quick_app/models/chi_tiet_hoa_don.dart';
import 'package:go_quick_app/models/nhan_vien.dart';
import 'package:go_quick_app/models/thanh_toan.dart';
import 'package:meta/meta.dart';
import 'dart:convert';

List<HoaDon> listHoaDonFromJson(String str) =>
    List<HoaDon>.from(json.decode(str).map((x) => HoaDon.fromJson(x)));

String listHoaDonToJson(List<HoaDon> data) =>
    json.encode(List<dynamic>.from(data.map((x) => x.toJson())));

class HoaDon {
  HoaDon({
    this.maHoaDon,
    this.nguoiLapHoaDon,
    this.tongThanhTien,
    this.thanhToan,
    this.ban,
    this.ghiChu,
    this.daTraHoaDon,
    this.tinhTrang,
    this.createdAt,
    this.updatedAt,
  });

  int? maHoaDon;
  NhanVien? nguoiLapHoaDon;
  double? tongThanhTien;
  ThanhToan? thanhToan;
  Ban? ban;
  String? ghiChu;
  bool? daTraHoaDon;
  String? tinhTrang;
  List<ChiTietHoaDon>? chiTietHoaDons;
  DateTime? createdAt;
  DateTime? updatedAt;

  factory HoaDon.fromJson(Map<String, dynamic> json) => HoaDon(
        maHoaDon: json["maHoaDon"],
        nguoiLapHoaDon: NhanVien.fromJson(json["nguoiLapHoaDon"]),
        tongThanhTien: json["tongThanhTien"],
        thanhToan: json["thanhToan"] != null
            ? ThanhToan.fromJson(json["thanhToan"])
            : null,
        ban: Ban.fromJson(json["ban"]),
        daTraHoaDon: json["daTraHoaDon"],
        tinhTrang: json["tinhTrang"],
        ghiChu: json["ghiChu"],
        createdAt: json["createdAt"] != null
            ? DateTime.parse(json["createdAt"]).add(const Duration(hours: 7))
            : null,
        updatedAt: json["updatedAt"] != null
            ? DateTime.parse(json["updatedAt"]).add(const Duration(hours: 7))
            : null,
      );

  Map<String, dynamic> toJson() => {
        "maHoaDon": maHoaDon,
        "nguoiLapHoaDon": nguoiLapHoaDon?.toJson(),
        // ignore: prefer_null_aware_operators
        "tongThanhTien": tongThanhTien,
        "thanhToan": thanhToan?.toJson(),
        "ghiChu": ghiChu,
        "ban": ban?.toJson(),
        "tinhTrang": tinhTrang,
        "daTraHoaDon": daTraHoaDon
      };
}
